### DISCUSSION TITLE
Nginx configuration with Docker, HTTPs

### DISCUSSION THREAD

**Discussion by 20v100** [coder/code-server#2653] [Q&A]
I'm able to get Code-Server working with Docker, NGINX and HTTPs. However, I don't know how to further modify my `app.conf` file to access code-sever via the endpoint `\code\`. Currently it serve Code-server at root `\` endpoint. Here is the relevant 443 section on of my `app.conf` file:

```
server {
    listen       443 ssl http2;
    listen       [::]:443 ssl http2;
    server_name xxx;

    ssl_certificate /etc/letsencrypt/live/xxx/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/xxx/privkey.pem;
    include /etc/letsencrypt/options-ssl-nginx.conf;
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;

    root                 /var/www/xxx/;
    index                index.html;
    autoindex            off;

    location /
    {
        try_files /nonexistent @$http_upgrade;
    }

    location @websocket	 
	  {
        # websocket related stuff
        proxy_pass           http://backend_code;    
        proxy_http_version   1.1;
        proxy_set_header     Upgrade $http_upgrade;
        proxy_set_header     Host $host;
        proxy_set_header     X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header     X-Forwarded-Proto $scheme;
        proxy_set_header     Connection "Upgrade";
        proxy_cache_bypass   $http_upgrade;
    }

    location @ 
	   {
         # web related stuff
         proxy_pass           http://backend_code;    
         proxy_http_version   1.1;
         proxy_set_header     Upgrade $http_upgrade;
         proxy_set_header     Host $host;
         proxy_set_header     X-Forwarded-For $proxy_add_x_forwarded_for;
         proxy_set_header     X-Forwarded-Proto $scheme;
         proxy_set_header     Connection keep-alive;
         proxy_cache_bypass   $http_upgrade;
    }

...
```

The below modification to the location results in a 404 page:

```
    location /code/
    {
        try_files /nonexistent @$http_upgrade;
    }
```

Thank you in advance for your help. Much appreciated!

**Comment by code-asher** [comment]
I think you might just need a trailing slash:

```
proxy_pass           http://backend_code/;
```

That causes NGINX to remove the base path (/code) from the URL before
proxying to to code-server.

**Reply by 20v100** [reply]
Thanks for the help. It did not work however. I get `nginx: [emerg] "proxy_pass" cannot have URI part in location given by regular expression, or inside named location, or inside "if" statement, or inside "limit_except" block in /etc/nginx/conf.d/app.conf:55` when adding the `/`.

I think the redirect to `backend_code` is working when I add the `location /code/` because I get a 404 page layout that is different than Nginx default 404. Hence, I think the 404 is served by Code-Server. So I may be missing a setup in Code-Server to enable it via the `\code\` endpoint?

**Reply by code-asher** [reply]
Ahhhh I see, I'm not familiar with this syntax so I'm not sure how exactly to make it work. Usually I use:

```
location /code {
  proxy_pass http://localhost:8080/; # < this trailing slash
  ... other proxy stuff
}
```

There isn't anything to set up on code-server's side; it expects the reverse proxy to rewrite `/code` to `/` before proxying the request. If it doesn't rewrite then code-server ends up getting `/code` and that's a 404 since that page doesn't exist.  The proxying itself is definitely working, we just need to make NGINX rewrite somehow.

**Reply by code-asher** [reply]
Potentially helpful links:

http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass
https://stackoverflow.com/a/22759570

They make it seem like it should work so it's definitely odd you're getting that error.
